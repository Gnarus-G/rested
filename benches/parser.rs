use criterion::{criterion_group, criterion_main, Criterion};
use rested::parser;

const SOURCE: &str = r#"
set BASE_URL env("base")

[env("token")]

let string = "death"

post /todos {
   header "Authorization" env("token")
   body json({
    userId: 1,
    id: 999,
    "title": `delectus aut ${string}`,
    "completed": env("base")
  })
}
"#;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse ast", |b| {
        b.iter(|| {
            let mut parser = parser::Parser::new(SOURCE);
            let _ = parser.parse();
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
