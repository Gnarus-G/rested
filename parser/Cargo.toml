[package]
name = "parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
colored = "2.0.0"
serde = { version = "1.0.159", features = ["derive"] }
lexer = { path = "../lexer" }
error_meta = { path = "../error_meta" }

[dev-dependencies]
insta = { version = "1.29.0", features = ["ron"] }
