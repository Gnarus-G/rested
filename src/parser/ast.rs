use std::fmt::Display;

use serde::Serialize;

use crate::{
    error_meta::ContextualError,
    lexer::{
        locations::{GetSpan, Position, Span},
        Token,
    },
    utils,
};

use self::result::ParsedNode;

use super::error::ParseError;

type Error<'source> = ContextualError<ParseError<'source>>;

#[derive(Debug, PartialEq, Serialize)]
pub struct Program<'i> {
    pub source: &'i str,
    pub items: utils::Array<Item<'i>>,
}

impl<'i> Program<'i> {
    pub fn new(source: &'i str, items: Vec<Item<'i>>) -> Self {
        Self {
            source,
            items: items.into(),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Literal<'i> {
    pub value: &'i str,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct StringLiteral<'source> {
    pub raw: &'source str,
    pub value: &'source str,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Block<'source> {
    pub statements: Vec<Statement<'source>>,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Item<'source> {
    Set {
        identifier: ParsedNode<'source, Token<'source>>,
        value: Expression<'source>,
    },
    Let {
        identifier: ParsedNode<'source, Token<'source>>,
        value: Expression<'source>,
    },
    LineComment(Literal<'source>),
    Request {
        method: RequestMethod,
        endpoint: Endpoint<'source>,
        block: Option<Block<'source>>,
        span: Span,
    },
    Expr(Expression<'source>),
    Attribute {
        location: Position,
        identifier: ParsedNode<'source, Token<'source>>,
        parameters: Option<Arguments<'source>>,
    },
    Error(Box<Error<'source>>),
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub enum RequestMethod {
    GET,
    POST,
    DELETE,
    PATCH,
    PUT,
}

impl Display for RequestMethod {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Statement<'i> {
    Header {
        name: ParsedNode<'i, StringLiteral<'i>>,
        value: Expression<'i>,
    },
    Body {
        value: Expression<'i>,
        start: Position,
    },
    LineComment(Literal<'i>),
    Error(Box<Error<'i>>),
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Arguments<'source> {
    pub span: Span,
    pub parameters: Vec<Expression<'source>>,
}

impl<'source> Arguments<'source> {
    pub fn iter(&self) -> impl Iterator<Item = &Expression<'source>> {
        self.parameters.iter()
    }
}

pub type Spanned<T> = (Span, T);

#[derive(Debug, PartialEq, Serialize)]
pub enum Expression<'source> {
    Identifier(ParsedNode<'source, Token<'source>>),
    String(StringLiteral<'source>),
    Bool(Literal<'source>),
    Number(Literal<'source>),
    Call {
        identifier: ParsedNode<'source, Token<'source>>,
        arguments: Arguments<'source>,
    },
    Array(Spanned<Vec<ArrayElement<'source>>>),
    Object(Spanned<Vec<ObjectEntry<'source>>>),
    Null(Span),
    EmptyArray(Span),
    EmptyObject(Span),
    TemplateSringLiteral {
        span: Span,
        parts: Vec<Expression<'source>>,
    },
    Error(Box<Error<'source>>),
}

#[derive(Debug, PartialEq, Serialize)]
pub struct ArrayElement<'source> {
    pub expr: Expression<'source>,
    // For missing commas actually. I don't know if this is a good idea generally though.
    pub errors: Vec<Error<'source>>,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct ObjectEntry<'source> {
    pub key: ParsedNode<'source, StringLiteral<'source>>,
    pub value: Expression<'source>,
    // For missing commas...
    pub errors: Vec<Error<'source>>,
}

impl<'source> ObjectEntry<'source> {
    pub fn new(
        key: ParsedNode<'source, StringLiteral<'source>>,
        value: Expression<'source>,
    ) -> Self {
        Self {
            key,
            value,
            errors: vec![],
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Endpoint<'i> {
    Url(Literal<'i>),
    Pathname(Literal<'i>),
}

pub mod result {

    use super::*;

    #[derive(Debug, PartialEq, serde::Serialize)]
    pub enum ParsedNode<'i, T: GetSpan> {
        Ok(T),
        Error(Box<Error<'i>>),
    }

    impl<'source, T: GetSpan> ParsedNode<'source, T> {
        pub fn get(&self) -> std::result::Result<&T, Box<Error<'source>>> {
            match self {
                ParsedNode::Ok(node) => Ok(node),
                ParsedNode::Error(error) => Err(error.clone()),
            }
        }
    }
}

mod convert {
    use super::*;

    impl<'i> From<&Token<'i>> for ParsedNode<'i, Token<'i>> {
        fn from(token: &Token<'i>) -> Self {
            Self::Ok(Token {
                kind: token.kind,
                text: token.text,
                start: token.start,
            })
        }
    }
    impl<'i> From<&Token<'i>> for Literal<'i> {
        fn from(token: &Token<'i>) -> Self {
            Self {
                value: token.text,
                span: token.span(),
            }
        }
    }
    impl<'i> From<&Token<'i>> for StringLiteral<'i> {
        fn from(token: &Token<'i>) -> Self {
            let value = match (token.text.chars().next(), token.text.chars().last()) {
                (Some('"'), Some('"')) if token.text.len() > 1 => {
                    &token.text[1..token.text.len() - 1]
                }
                (Some('`'), Some('`')) if token.text.len() > 1 => {
                    &token.text[1..token.text.len() - 1]
                }
                (_, Some('`')) => &token.text[..token.text.len() - 1],
                (Some('`'), _) => &token.text[1..],
                _ => token.text,
            };

            Self {
                raw: token.text,
                value,
                span: token.span(),
            }
        }
    }

    impl<'source, T: GetSpan> From<std::result::Result<T, Box<Error<'source>>>>
        for ParsedNode<'source, T>
    {
        fn from(value: std::result::Result<T, std::boxed::Box<Error<'source>>>) -> Self {
            match value {
                Ok(value) => ParsedNode::Ok(value),
                Err(error) => ParsedNode::Error(error),
            }
        }
    }

    impl<'source> From<Expression<'source>> for ArrayElement<'source> {
        fn from(value: Expression<'source>) -> Self {
            Self {
                expr: value,
                errors: vec![],
            }
        }
    }

    impl<'source> From<Error<'source>> for Expression<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<Error<'source>> for Statement<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<Box<Error<'source>>> for Statement<'source> {
        fn from(value: Box<Error<'source>>) -> Self {
            Self::Error(value)
        }
    }

    impl<'source> From<Error<'source>> for Item<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<std::result::Result<Self, Box<Error<'source>>>> for Expression<'source> {
        fn from(value: std::result::Result<Self, Box<Error<'source>>>) -> Self {
            match value {
                Ok(v) => v,
                Err(error) => Self::Error(error),
            }
        }
    }
}
