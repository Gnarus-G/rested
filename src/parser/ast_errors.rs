use super::{
    ast::{self, Expression, Item, Statement},
    error::ParseError,
};
use crate::error_meta::ContextualError;

pub trait GetErrors<'source> {
    fn errors(&self) -> Vec<ContextualError<ParseError<'source>>>;
}

impl<'source> GetErrors<'source> for Item<'source> {
    fn errors(&self) -> Vec<ContextualError<ParseError<'source>>> {
        let mut errors = vec![];

        match &self {
            Item::Set { identifier, value } => {
                if let ast::result::ParsedNode::Error(error) = identifier {
                    errors.push(*error.clone());
                }
                errors.extend(value.errors());
            }
            Item::Let { value, identifier } => {
                if let ast::result::ParsedNode::Error(error) = identifier {
                    errors.push(*error.clone());
                }
                errors.extend(value.errors())
            }
            Item::LineComment(_) => {}
            Item::Request {
                block: Some(block), ..
            } => errors.extend(block.statements.iter().flat_map(|s| s.errors())),
            Item::Expr(expr) => errors.extend(expr.errors()),
            Item::Attribute { parameters, .. } => {
                if let Some(args) = parameters.as_ref() {
                    errors.extend(args.parameters.iter().flat_map(|expr| expr.errors()))
                }
            }
            Item::Error(e) => errors.push(*e.clone()),
            _ => {}
        }

        errors
    }
}

impl<'source> GetErrors<'source> for Statement<'source> {
    fn errors(&self) -> Vec<ContextualError<ParseError<'source>>> {
        let mut errors = vec![];

        match &self {
            Statement::Header { value, name } => {
                if let ast::result::ParsedNode::Error(error) = name {
                    errors.push(*error.clone());
                }
                errors.extend(value.errors())
            }
            Statement::Body { value, .. } => errors.extend(value.errors()),
            Statement::LineComment(_) => {}
            Statement::Error(e) => errors.push(*e.clone()),
        }

        errors
    }
}

impl<'source> GetErrors<'source> for Expression<'source> {
    fn errors(&self) -> Vec<ContextualError<ParseError<'source>>> {
        let mut errors = vec![];
        match &self {
            Expression::Call { arguments, .. } => {
                errors.extend(arguments.iter().flat_map(|e| e.errors()))
            }
            Expression::Array((_, arr)) => errors.extend(arr.iter().flat_map(|e| e.errors())),
            Expression::Object((_, o)) => errors.extend(o.values().flat_map(|e| e.errors())),
            Expression::TemplateSringLiteral { parts, .. } => {
                parts.iter().for_each(|expr| errors.extend(expr.errors()))
            }
            Expression::Error(e) => errors.push(*e.clone()),
            _ => {}
        }

        errors
    }
}
