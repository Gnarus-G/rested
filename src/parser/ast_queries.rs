use super::{
    ast::{Identifier, Program},
    ast_errors::GetErrors,
    error::ParseError,
};
use crate::{
    error_meta::ContextualError,
    lexer::{locations::Location, Array},
};

impl<'source> Program<'source> {
    pub fn variables(&self) -> impl Iterator<Item = &Identifier<'source>> {
        self.items.iter().filter_map(|i| match i {
            super::ast::Item::Let { identifier, .. } => Some(identifier),
            _ => None,
        })
    }

    pub fn variables_before(&self, location: Location) -> Array<&Identifier<'source>> {
        self.variables()
            .filter(|i| Into::<Location>::into(i.span.start).is_before(location))
            .collect()
    }

    pub fn errors(&self) -> Vec<ContextualError<ParseError<'source>>> {
        let mut errors = vec![];
        for item in &self.items {
            errors.extend(item.errors())
        }

        errors
    }
}
