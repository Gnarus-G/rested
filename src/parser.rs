use crate::{
    ast::{Header, Program, Request},
    error::{ParseError, ParseErrorKind},
    lexer::{Lexer, Token, TokenKind},
};

type Result<T> = std::result::Result<T, ParseError>;

#[derive(Debug)]
pub struct Parser<'a> {
    lexer: Lexer<'a>,
    peeked: Option<Token<'a>>,
}

impl<'i> Parser<'i> {
    pub fn new(lexer: Lexer<'i>) -> Self {
        Self {
            lexer,
            peeked: None,
        }
    }

    fn token(&mut self) -> Token<'i> {
        match self.peeked.take() {
            Some(t) => t,
            None => self.lexer.next(),
        }
    }

    fn peek_token(&mut self) -> &Token<'i> {
        self.peeked.get_or_insert_with(|| self.lexer.next())
    }

    fn eat_token(&mut self) {
        self.token();
    }

    pub fn parse(&mut self) -> Result<Program<'i>> {
        let mut program = Program::new();

        use crate::lexer::TokenKind::*;

        let mut token = self.token();

        while token.kind != End {
            let request = match token.kind {
                Get => self.parse_get_request()?,
                tk => todo!("{tk:?}"),
            };
            program.requests.push(request);
            token = self.token();
        }

        Ok(program)
    }

    fn parse_get_request(&mut self) -> Result<Request<'i>> {
        Ok(Request::Get(crate::ast::GetRequestParams {
            url: self.token().text,
            headers: self.parse_get_params()?,
        }))
    }

    fn parse_get_params(&mut self) -> Result<Option<Vec<Header<'i>>>> {
        let mut token = self.token();
        if let crate::lexer::TokenKind::LBracket = token.kind {
            token = self.token();
            let mut headers = vec![];
            while token.kind != TokenKind::RBracket {
                let h = match token.kind {
                    TokenKind::Header => self.parse_header()?,
                    TokenKind::Ident => todo!(),
                    TokenKind::StringLiteral => todo!(),
                    TokenKind::Assign => todo!(),
                    TokenKind::Quote => todo!(),
                    tk => todo!("{tk:?}"),
                };

                headers.push(h);
                token = self.token();
            }
            Ok(Some(headers))
        } else {
            Ok(None)
        }
    }

    fn parse_header(&mut self) -> Result<Header<'i>> {
        let t = self.token();

        let header_name = t.text;

        self.expect(TokenKind::Assign)?;

        self.eat_token();

        self.expect(TokenKind::Quote)?;

        self.eat_token();

        self.expect(TokenKind::StringLiteral)?;

        let header_value = self.token().text;

        self.expect(TokenKind::Quote)?;

        self.eat_token();

        Ok(Header {
            name: header_name,
            value: header_value,
        })
    }

    fn expect(&mut self, kind: TokenKind) -> Result<()> {
        let ahead = self.peek_token();

        if ahead.kind == kind {
            return Ok(());
        }

        Err(ParseError::new(
            ParseErrorKind::Unexpected {
                expected: kind,
                found: ahead.kind,
            },
            ahead.location,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::ast::{GetRequestParams, Header, Program, Request};

    macro_rules! assert_program {
        ($input:literal, $program:expr) => {
            let lexer = Lexer::new($input);
            let mut parser = Parser::new(lexer);
            assert_eq!(parser.parse().unwrap(), $program);
        };
    }

    #[test]
    fn parse_get_url() {
        assert_program!(
            "get http://localhost",
            Program {
                requests: vec![Request::Get(GetRequestParams {
                    url: "http://localhost",
                    headers: None
                })]
            }
        );
    }

    #[test]
    fn parse_get_with_headers() {
        assert_program!(
            "get http://localhost { header Authorization = \"Bearer token\" }",
            Program {
                requests: vec![Request::Get(GetRequestParams {
                    url: "http://localhost",
                    headers: Some(vec![Header {
                        name: "Authorization",
                        value: "Bearer token"
                    }])
                })]
            }
        );
    }
}
