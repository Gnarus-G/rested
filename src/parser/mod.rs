pub mod ast;
mod ast_errors;
mod ast_queries;
mod ast_span;
pub mod error;

use std::collections::BTreeMap;

use ast::{Endpoint, Expression, Item, RequestMethod, Statement};

use self::ast::{Arguments, Block};
use self::error::{Expectations, ParseError};

use crate::error_meta::ContextualError;
use crate::lexer::locations::{GetSpan, Position, Span};
use crate::lexer::TokenKind;
use crate::lexer::TokenKind::*;
use crate::lexer::{Lexer, Token};

macro_rules! match_or_throw {
    ($expression:expr; $expectations:ident; $self:ident; $( $( $pattern:ident )|+ $( if $guard: expr )? => $arm:expr $(,)? )+ $( ,$message:literal )? ) => {
        match $expression {
            $(
                $( TokenKind::$pattern )|+ $( if $guard )? => $arm,
            )+
            _ => return Err($expectations.expected_one_of_tokens($self.curr_token(), &[$( $( $pattern ),+ ),+])$(.with_message($message))?.into())
        }
    };
}

pub type Result<'source, T> = std::result::Result<T, ContextualError<ParseError<'source>>>;

trait TokenCheck {
    fn is_one_of(&self, kinds: &[TokenKind]) -> bool;
    fn is(&self, kind: TokenKind) -> bool;
}

impl<'source> TokenCheck for Token<'source> {
    fn is_one_of(&self, kinds: &[TokenKind]) -> bool {
        kinds.contains(&self.kind)
    }

    fn is(&self, kind: TokenKind) -> bool {
        self.kind == kind
    }
}

#[derive(Debug)]
pub struct Parser<'i> {
    lexer: Lexer<'i>,
    token: Option<Token<'i>>,
    peeked: Option<Token<'i>>,
}

impl<'i> Parser<'i> {
    pub fn new(code: &'i str) -> Self {
        Self {
            peeked: None,
            lexer: Lexer::new(code),
            token: None,
        }
    }

    fn curr_token(&self) -> &Token<'i> {
        self.token
            .as_ref()
            .expect("self.token should be initialized at the start of parsing")
    }

    fn next_token(&mut self) -> &Token<'i> {
        self.token = match self.peeked.take() {
            Some(t) => Some(t),
            None => Some(self.lexer.next_token()),
        };
        self.curr_token()
    }

    fn peek_token(&mut self) -> &Token<'i> {
        self.peeked.get_or_insert_with(|| self.lexer.next_token())
    }

    fn eat_till_next_top_level_peek_token(&mut self) {
        loop {
            let is_top_level_token_ahead = matches!(
                self.peek_token().kind,
                Get | Post | Put | Patch | Delete | Set | AttributePrefix | Let | End
            );

            if is_top_level_token_ahead {
                break;
            }

            self.next_token();
        }
    }

    fn span_from(&self, start: Position) -> Span {
        start.to_end_of(self.curr_token().span())
    }

    pub fn parse(&mut self) -> ast::Program<'i> {
        let mut items: Vec<_> = vec![];

        use crate::lexer::TokenKind::*;

        self.next_token();

        while self.curr_token().kind != End {
            let result: std::result::Result<ast::Item<'_>, _> = match self.curr_token().kind {
                Get => self.parse_request(RequestMethod::GET),
                Post => self.parse_request(RequestMethod::POST),
                Put => self.parse_request(RequestMethod::PUT),
                Patch => self.parse_request(RequestMethod::PATCH),
                Delete => self.parse_request(RequestMethod::DELETE),
                Linecomment | Shebang => Ok(Item::LineComment(self.curr_token().into())),
                Set => self.parse_set_statement(),
                AttributePrefix => {
                    let e = Expectations::new(self);
                    let item = self.parse_attribute();

                    if item.is_ok() {
                        let valid_after_attribute =
                            [Get, Post, Put, Patch, Delete, AttributePrefix, Linecomment];

                        if let Err(err) = e.expect_peek_one_of(self, &valid_after_attribute) {
                            items.push(
                                err.with_message(
                                    "after attributes should come requests or more attributes",
                                )
                                .into(),
                            );
                            continue;
                        }
                    }

                    item
                }
                Let => self.parse_let_statement(),
                _ => match self.parse_expression() {
                    Ok(exp) => Ok(Item::Expr(exp)),
                    Err(err) => Err(err),
                },
            };

            match result {
                Ok(item) => items.push(item),
                Err(error) => {
                    items.push(error.into());
                    self.eat_till_next_top_level_peek_token();
                }
            }

            self.next_token();
        }

        return ast::Program { items };
    }

    fn parse_request(&mut self, method: RequestMethod) -> Result<'i, Item<'i>> {
        let e = Expectations::new(self);

        let url = self.next_token();

        let endpoint = match_or_throw! { url.kind; e; self;
            Url => Endpoint::Url(url.into()),
            Pathname => Endpoint::Pathname(url.into()),
            "expecting only a url and pathname here"
        };

        let url_span: Span = url.span();

        let block = self.parse_block();

        let span_next = if let Some(b) = block.as_ref() {
            b.span
        } else {
            url_span
        };

        Ok(Item::Request {
            span: e.start.to_end_of(span_next),
            method,
            endpoint,
            block,
        })
    }

    fn parse_set_statement(&mut self) -> Result<'i, Item<'i>> {
        let e = Expectations::new(self);
        let identifier = match e.expect_peek(self, TokenKind::Ident) {
            Ok(i) => i.into(),
            Err(error) => {
                return Ok(Item::Set {
                    identifier: ast::TokenNode::Error(error.clone().into()),
                    value: Expression::Error(error),
                })
            }
        };

        self.next_token();

        Ok(Item::Set {
            value: match self.parse_expression() {
                Ok(expr) => expr,
                Err(error) => {
                    return Ok(Item::Set {
                        identifier,
                        value: Expression::Error(error),
                    })
                }
            },
            identifier,
        })
    }

    fn parse_block(&mut self) -> Option<Block<'i>> {
        let LBracket = self.peek_token().kind else {
            return None;
        };

        let span_start = self.next_token().start; // remember LBracket's location
        self.next_token();
        let mut statements: Vec<Statement<'i>> = vec![];

        while self.curr_token().kind != RBracket && self.curr_token().kind != End {
            let statement = match self.parse_statement() {
                Ok(s) => s,
                Err(error) => error.into(),
            };
            statements.push(statement);
            self.next_token();
        }

        return Some(Block {
            statements,
            span: Span::new(span_start, self.curr_token().start), // span to RBracket's location
        });
    }

    fn parse_statement(&mut self) -> Result<'i, Statement<'i>> {
        let e = Expectations::new(self);

        let statement = match_or_throw! { self.curr_token().kind; e; self;
            Header => self.parse_header()?,
            Body => self.parse_body()?,
            Linecomment | Shebang => Statement::LineComment(self.curr_token().into()),
            "may only declare headers or a body statement here"
        };

        Ok(statement)
    }

    fn parse_header(&mut self) -> Result<'i, Statement<'i>> {
        let e = Expectations::new(self);
        let header_name = e
            .expect_peek(self, TokenKind::StringLiteral)
            .map(|t| t.into())
            .into();

        self.next_token();

        let value = match self.parse_expression() {
            Ok(e) => e,
            Err(error) => {
                return Ok(Statement::Header {
                    name: header_name,
                    value: Expression::Error(error),
                })
            }
        };

        Ok(Statement::Header {
            name: header_name,
            value,
        })
    }

    fn parse_body(&mut self) -> Result<'i, Statement<'i>> {
        let start = self.curr_token().start;

        self.next_token();

        let value = match self.parse_expression() {
            Ok(e) => e,
            Err(error) => {
                return Ok(Statement::Body {
                    value: Expression::Error(error),
                    start,
                })
            }
        };

        Ok(Statement::Body { value, start })
    }

    fn parse_expression(&mut self) -> Result<'i, Expression<'i>> {
        let e = Expectations::new(self);
        let kind = self.curr_token().kind;

        let exp = match_or_throw! { kind; e; self;
            Ident if self.peek_token().kind == LParen => self.parse_call_expression()?,
            Ident => Expression::Identifier(self.curr_token().into()),
            StringLiteral => Expression::String(self.curr_token().into()),
            Boolean => Expression::Bool(self.curr_token().into()),
            Number => Expression::Number(self.curr_token().into()),
            MultiLineStringLiteral => self.parse_multiline_string_literal()?,
            LBracket | LSquare => self.parse_json_like()?,
            Null => Expression::Null(self.curr_token().span()),
        };

        Ok(exp)
    }

    fn parse_json_like(&mut self) -> Result<'i, Expression<'i>> {
        let start_token = self.curr_token();
        let e = Expectations::new(self);

        let object = match start_token.kind {
            LBracket => {
                if self.peek_token().kind == RBracket {
                    self.next_token();
                    return Ok(Expression::EmptyObject(self.span_from(e.start)));
                }

                let mut fields = BTreeMap::new();

                while self.peek_token().kind != RBracket {
                    if self.peek_token().is(Linecomment) {
                        self.next_token();
                        continue;
                    }

                    let (key, value) = self.parse_object_property()?;
                    fields.insert(key, value);

                    if !self.peek_token().is(RBracket) {
                        e.expect_peek(self, Comma)?;
                    }
                }

                self.next_token();

                let span = e.start.to_end_of(self.curr_token().span());

                Expression::Object((span, fields))
            }
            LSquare => {
                if self.peek_token().kind == RSquare {
                    self.next_token();
                    return Ok(Expression::EmptyArray(self.span_from(e.start)));
                }

                let mut list = vec![];

                while self.peek_token().kind != RSquare {
                    if self.peek_token().is(Linecomment) {
                        self.next_token();
                        continue;
                    }

                    self.next_token();
                    list.push(self.parse_json_like()?);

                    if !self.peek_token().is(RSquare) {
                        e.expect_peek(self, Comma)?;
                    }
                }

                self.next_token();

                let span = e.start.to_end_of(self.curr_token().span());

                Expression::Array((span, list))
            }
            _ => self.parse_expression()?,
        };

        Ok(object)
    }

    fn parse_object_property(&mut self) -> Result<'i, (&'i str, Expression<'i>)> {
        let e = Expectations::new(self);
        let key_token = self.next_token();

        let key = match_or_throw! { key_token.kind; e; self;
            Ident => key_token.text,
            StringLiteral => ast::StringLiteral::from(key_token).value
        };

        e.expect_peek(self, TokenKind::Colon)?;
        self.next_token();

        Ok((key, self.parse_json_like()?))
    }

    fn parse_call_expression(&mut self) -> Result<'i, Expression<'i>> {
        let identifier = self.curr_token().into();
        self.next_token();

        self.next_token();

        let mut arguments = vec![];

        while self.curr_token().kind != TokenKind::RParen {
            let exp = self.parse_expression()?;
            arguments.push(exp);
            self.next_token();
        }

        Ok(Expression::Call {
            identifier,
            arguments,
        })
    }

    fn parse_multiline_string_literal(&mut self) -> Result<'i, Expression<'i>> {
        let mut parts = vec![];
        let e = Expectations::new(self);
        let end;

        loop {
            let kind = self.curr_token().kind;

            match_or_throw! { kind; e; self;
                MultiLineStringLiteral
                    if self.peek_token().kind == TokenKind::DollarSignLBracket =>
                {
                    let s_literal = Expression::String(self.curr_token().into());

                    parts.push(s_literal);

                    self.next_token();

                    self.next_token();

                    parts.push(self.parse_expression()?);
                }
                End => {
                    end = self.curr_token().end_position();
                    break;
                },
                MultiLineStringLiteral if parts.is_empty() => {
                    return Ok(Expression::String(self.curr_token().into()));
                }
                MultiLineStringLiteral => {
                    end = self.curr_token().end_position();
                    parts.push(Expression::String(self.curr_token().into()));
                    break;
                }
            };

            self.next_token();
        }

        Ok(Expression::TemplateSringLiteral {
            span: Span::new(e.start, end),
            parts,
        })
    }

    fn parse_attribute(&mut self) -> Result<'i, Item<'i>> {
        let e = Expectations::new(self);

        let identifier = e.expect_peek(self, TokenKind::Ident)?.into();

        if self.peek_token().kind != TokenKind::LParen {
            return Ok(Item::Attribute {
                location: e.start,
                identifier,
                parameters: None,
            });
        }

        self.next_token();

        Ok(Item::Attribute {
            location: e.start,
            identifier,
            parameters: Some(self.parse_parameters()),
        })
    }

    fn parse_parameters(&mut self) -> Arguments<'i> {
        let mut params = vec![];
        let params_start = self.curr_token().start;

        self.next_token();

        while self.curr_token().kind != TokenKind::RParen
            && self.curr_token().kind != TokenKind::End
        {
            let exp = match self.parse_expression() {
                Ok(exp) => exp,
                Err(error) => Expression::Error(error),
            };

            params.push(exp);

            self.next_token();
        }

        let last_token = self.curr_token(); // should be RParen

        Arguments {
            span: Span {
                start: params_start,
                end: last_token.span().end,
            },
            parameters: params,
        }
    }

    fn parse_let_statement(&mut self) -> Result<'i, Item<'i>> {
        let e = Expectations::new(self);
        let identifier = self.next_token().into();

        e.expect_peek(self, TokenKind::Assign)?;

        self.next_token();

        Ok(Item::Let {
            value: match self.parse_expression() {
                Ok(e) => e,
                Err(error) => {
                    return Ok(Item::Let {
                        identifier,
                        value: Expression::Error(error),
                    })
                }
            },
            identifier,
        })
    }
}
