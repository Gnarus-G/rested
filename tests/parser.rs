use insta::assert_ron_snapshot;

use rested::parser::Parser;

#[test]
fn it_works() {
    let code = r#"
set BASE_URL "httas..."
post http://lasdf.. {}
// asdfasdf

let output_file = "output/file.json"
let token = env("auth-token")

@log(output_file)
get /asd {
  // asdfasd
  header "Authorization" token
  body `{
      "neet": "${env("var")}",
      "nerd": "${output_file}",
  }`
}"#;

    let p = Parser::new(code).parse().unwrap();
    insta::with_settings!({
         description => code
    }, {
        assert_ron_snapshot!(p);
    })
}
