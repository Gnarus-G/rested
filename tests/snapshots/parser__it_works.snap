---
source: tests/parser.rs
description: "\nset BASE_URL \"httas...\"\npost http://lasdf.. {}\n// asdfasdf\n\nlet output_file = \"output/file.json\"\nlet token = env(\"auth-token\")\n\n@log(output_file)\nget /asd {\n  // asdfasd\n  header \"Authorization\" token\n  body `{\n      \"neet\": \"${env(\"var\")}\",\n      \"nerd\": \"${output_file}\",\n  }`\n}"
expression: p
---
Program(
  items: [
    Set(
      identifier: Identifier(
        name: "BASE_URL",
        location: Location(
          line: 1,
          col: 4,
        ),
      ),
      value: String(Literal(
        value: "httas...",
        location: Location(
          line: 1,
          col: 13,
        ),
      )),
    ),
    Request(
      method: POST,
      endpoint: Url(Literal(
        value: "http://lasdf..",
        location: Location(
          line: 2,
          col: 5,
        ),
      )),
      params: [],
      location: Location(
        line: 2,
        col: 0,
      ),
    ),
    LineComment(Literal(
      value: "// asdfasdf",
      location: Location(
        line: 3,
        col: 0,
      ),
    )),
    Let(
      identifier: Identifier(
        name: "output_file",
        location: Location(
          line: 5,
          col: 4,
        ),
      ),
      value: String(Literal(
        value: "output/file.json",
        location: Location(
          line: 5,
          col: 18,
        ),
      )),
    ),
    Let(
      identifier: Identifier(
        name: "token",
        location: Location(
          line: 6,
          col: 4,
        ),
      ),
      value: Call(
        identifier: Identifier(
          name: "env",
          location: Location(
            line: 6,
            col: 12,
          ),
        ),
        arguments: [
          String(Literal(
            value: "auth-token",
            location: Location(
              line: 6,
              col: 16,
            ),
          )),
        ],
      ),
    ),
    Attribute(
      location: Location(
        line: 8,
        col: 0,
      ),
      identifier: Identifier(
        name: "log",
        location: Location(
          line: 8,
          col: 1,
        ),
      ),
      parameters: [
        Identifier(Identifier(
          name: "output_file",
          location: Location(
            line: 8,
            col: 5,
          ),
        )),
      ],
    ),
    Request(
      method: GET,
      endpoint: Pathname(Literal(
        value: "/asd",
        location: Location(
          line: 9,
          col: 4,
        ),
      )),
      params: [
        LineComment(Literal(
          value: "// asdfasd",
          location: Location(
            line: 10,
            col: 2,
          ),
        )),
        Header(
          name: Literal(
            value: "Authorization",
            location: Location(
              line: 11,
              col: 9,
            ),
          ),
          value: Identifier(Identifier(
            name: "token",
            location: Location(
              line: 11,
              col: 25,
            ),
          )),
        ),
        Body(
          value: TemplateSringLiteral(
            parts: [
              String(Literal(
                value: "{\n      \"neet\": \"",
                location: Location(
                  line: 12,
                  col: 7,
                ),
              )),
              Call(
                identifier: Identifier(
                  name: "env",
                  location: Location(
                    line: 13,
                    col: 17,
                  ),
                ),
                arguments: [
                  String(Literal(
                    value: "var",
                    location: Location(
                      line: 13,
                      col: 21,
                    ),
                  )),
                ],
              ),
              String(Literal(
                value: "\",\n      \"nerd\": \"",
                location: Location(
                  line: 13,
                  col: 27,
                ),
              )),
              Identifier(Identifier(
                name: "output_file",
                location: Location(
                  line: 14,
                  col: 17,
                ),
              )),
              String(Literal(
                value: "\",\n  }",
                location: Location(
                  line: 14,
                  col: 28,
                ),
              )),
            ],
          ),
          location: Location(
            line: 12,
            col: 2,
          ),
        ),
      ],
      location: Location(
        line: 9,
        col: 0,
      ),
    ),
  ],
)
